src/hotspot/share/gc/shenandoah/shenandoahVMOperations.cpp

```diff
@@ -58,6 +58,10 @@ void VM_ShenandoahFinalMarkStartEvac::doit() {
   _gc->entry_final_mark();
 }
 
+void VM_ShenandoahFinalMarkStartEvac::doit_epilogue() {
+  OopMapCache::cleanup_old_entries();
+}
+
```

Comment:
Do you think, VM_ShenandoahFinalMarkStartEvac walks the stack roots.  If yes, i recommend adding OopMapCache::cleanup_old_entries() in  VM_ShenandoahOperation::doit_epilogue().  And this would made the change simple and also revert the change in this [PR](https://github.com/openjdk/jdk/pull/15921)

Reply:
No, `VM_ShenandoahFinalMarkStartEvac ` does not walk the stack roots, it signals the end of mark phase.

Reply:
Note we don't need to call   OopMapCache::cleanup_old_entries() after this [PR](https://github.com/openjdk/jdk/pull/15921).  VM_ShenandoahFinalMarkStartEvac is derived from VM_ShenandoahReferenceOperation. VM_ShenandoahReferenceOperation::doit_epilogue calls OopMapCache::cleanup_old_entries();